core/go/internal/sequencer/coordinator/coordinator.go

```diff
@@ -0,0 +1,600 @@
+		}
+	} else if c.domainAPI.ContractConfig().GetCoordinatorSelection() == prototk.ContractConfig_COORDINATOR_ENDORSER {
+		// E.g. Pente
+		// Make a fair choice about the next coordinator, but for now just choose the node this request arrived at
+		if len(c.originatorNodePool) == 0 {
+			log.L(ctx).Warnf("no pool to select a coordinator from yet")
+			return "", nil
+		} else {
+			// Round block number down to the nearest block range (e.g. block 1012, 1013, 1014 etc. all become 1000 for hashing)
+			effectiveBlockNumber := c.currentBlockHeight - (c.currentBlockHeight % c.coordinatorSelectionBlockRange)
+
+			// Take a numeric hash of the identities using the current block range
+			h := fnv.New32a()
+			h.Write([]byte(strconv.FormatUint(effectiveBlockNumber, 10)))
+			coordinator = c.originatorNodePool[int(h.Sum32())%len(c.originatorNodePool)]
+			log.L(ctx).Debugf("coordinator %s selected based on hash modulus of the originator pool", coordinator)
+		}
+	} else if c.domainAPI.ContractConfig().GetCoordinatorSelection() == prototk.ContractConfig_COORDINATOR_SENDER {
+		// E.g. Zeto
+		log.L(ctx).Debugf("coordinator %s selected as next active coordinator in originator coordinator mode", c.nodeName)
+		coordinator = c.nodeName
+	}
+
+	log.L(ctx).Debugf("selected active coordinator for contract %s: %s", c.contractAddress.String(), coordinator)
+
+	if strings.Contains(coordinator, "@") || coordinator == "" {
+		return "", i18n.NewError(ctx, msgs.MsgSequencerInternalError, "coordinator node is invalid")
+	}
+
+	return coordinator, nil
+}
+
+func (c *coordinator) UpdateOriginatorNodePool(ctx context.Context, originatorNode string) {
+	if strings.Contains(originatorNode, "@") {
+		log.L(ctx).Errorf("originator ID provided, originator node expected: %s", originatorNode)
+	}
+	log.L(ctx).Debugf("updating originator node pool with node %s", originatorNode)
+	if !slices.Contains(c.originatorNodePool, originatorNode) {
+		c.originatorNodePool = append(c.originatorNodePool, originatorNode)
+		slices.Sort(c.originatorNodePool)
+	}
+}
+
+// TODO consider renaming to setDelegatedTransactionsForOriginator to make it clear that we expect originators to include all inflight transactions in every delegation request and therefore this is
+// a replace, not an add.  Need to finalize the decision about whether we expect the originator to include all inflight delegated transactions in every delegation request. Currently the code assumes we do so need to make the spec clear on that point and
+// record a decision record to explain why.  Every  time we come back to this point, we will be tempted to reverse that decision so we need to make sure we have a record of the known consequences.
+// originator must be a fully qualified identity locator otherwise an error will be returned
+func (c *coordinator) addToDelegatedTransactions(ctx context.Context, originator string, transactions []*components.PrivateTransaction) error {
```

Comment:
Is this the whole list of full `PrivateTransaction` object that needs to be passed around? ... or surely just the TXIDs and originator information are passed.